### PR TITLE
migrate post and corresponding components to ts

### PR DESCRIPTION
**Pull request by dudupopkhadze**
#### Summary
Migrated  components/post_view/post/ index.js and post.jsx to typescript
also migrated corresponding test

Some notable points:

- Had to use `Any` type here, because could not infer  type from component created with  `injectIntl`

- Modified following compoonents because props that are passed with redux-connect function were marked as Proptype.isRequired and because of this  typescript compiler was throwing errors  

1.  components/post_view/post_body/post_body.jsx
2. components/post_view/post_header/post_header.jsx
3. components/post_profile_picture/post_profile_picture.jsx

All tests passing - No ts errors

#### Ticket Link

Fixes: https://github.com/mattermost/mattermost-server/issues/15464
JIRA:  https://mattermost.atlassian.net/browse/MM-20477

### CODE REVIEW

**Review by mkraft — APPROVED** [review]

File: components/post_view/post/index.ts [hunk 1]
```diff
@@ -36,12 +44,18 @@ export function isFirstReply(post, previousPost) {
...
 
+type OwnProps = {
+    post?: Post;
+    postId: string
+    previousPostId?: string;
+}
+
 function makeMapStateToProps() {
-    const getReplyCount = makeGetReplyCount();
+    const getReplyCount = makeGetReplyCount() as ((sate: GlobalState, post: Post) => number);
```
Comment (mickmister) [inline components/post_view/post/index.ts:54]:
```suggestion
    const getReplyCount = makeGetReplyCount() as ((state: GlobalState, post: Post) => number);
```

File: components/post_view/post/index.ts [hunk 2]
```diff
@@ -36,12 +44,18 @@ export function isFirstReply(post, previousPost) {
...
     const isPostCommentMention = makeIsPostCommentMention();
-    const createAriaLabelForPost = makeCreateAriaLabelForPost();
+    const createAriaLabelForPost = makeCreateAriaLabelForPost() as ((sate: GlobalState, post: Post) => (...args: any[]) => any);
```
Comment (mickmister) [inline components/post_view/post/index.ts:56]:
```suggestion
    const createAriaLabelForPost = makeCreateAriaLabelForPost() as ((state: GlobalState, post: Post) => (...args: any[]) => any);
```

File: components/post_view/post/post.tsx [hunk 3]
```diff
@@ -1,113 +1,122 @@
...
          */
-        isCommentMention: PropTypes.bool,
+    isCommentMention?: boolean,
 
-        /**
+    /**
          * The number of replies in the same thread as this post
          */
-        replyCount: PropTypes.number,
+    replyCount: number,
```
Comment (mickmister) [inline components/post_view/post/post.tsx:80]: This prop went from optional to required. Is this intentional?

File: components/post_view/post/post.tsx [hunk 4]
```diff
@@ -1,113 +1,122 @@
...
          */
-        previousPostIsComment: PropTypes.bool,
+    previousPostIsComment?: boolean,
 
-        /*
+    /*
          * Function called when the post options dropdown is opened
          */
-        togglePostMenu: PropTypes.func,
+    togglePostMenu: (value: boolean) => void,
```
Comment (mickmister) [inline components/post_view/post/post.tsx:70]: This prop also went from optional to required

File: components/post_view/post/post.tsx [hunk 5]
```diff
@@ -1,113 +1,122 @@
...
          */
-        isLastPost: PropTypes.bool,
+    isLastPost?: boolean,
 
-        /**
+    /**
          * Whether or not the channel that contains this post is archived
          */
-        channelIsArchived: PropTypes.bool.isRequired,
+    channelIsArchived?: boolean,
```
Comment (mickmister) [inline components/post_view/post/post.tsx:90]: This prop went from required to optional. Is this intentional?

File: components/post_view/post/post.tsx [hunk 6]
```diff
@@ -1,113 +1,122 @@
...
+    alt: boolean,
+    a11yActive: boolean,
+    currentAriaLabel: string,
+    ariaHidden: boolean,
+    fadeOutHighlight: boolean,
+}
 
-    static defaultProps = {
-        post: {},
-    };
```
Comment (mickmister) [inline components/post_view/post/post.tsx:108]: Why do we not need the `defaultProps` anymore?

File: components/post_view/post/post.tsx [hunk 7]
```diff
@@ -149,17 +158,19 @@ class Post extends React.PureComponent {
...
-        if (this.state.a11yActive) {
+        if (this.state.a11yActive && this.postRef.current) {
             this.postRef.current.dispatchEvent(new Event(A11yCustomEventTypes.UPDATE));
         }
     }
 
-    handleCommentClick = (e) => {
-        e.preventDefault();
+    handleCommentClick = () => {
+        // e.preventDefault();
```
Comment (mickmister) [inline components/post_view/post/post.tsx:173]: Can we remove this commented out code if it's unnecessary?

File: components/post_view/post/post.tsx [hunk 8]
```diff
@@ -168,20 +179,20 @@ class Post extends React.PureComponent {
...
 
-    handlePostClick = (e) => {
+    handlePostClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
         const post = this.props.post;
         if (!post) {
             return;
         }
 
-        if (this.props.channelIsArchived || post.system_post_ids) {
+        if (this.props.channelIsArchived) {
```
Comment (mickmister) [inline components/post_view/post/post.tsx:195]: Why did the conditional logic here change to not include `post.system_post_ids`?

File: components/post_view/post/post.tsx [hunk 9]
```diff
@@ -283,7 +294,7 @@ class Post extends React.PureComponent {
             className += ' post--pinned-or-flagged';
         }
 
-        if (this.state.alt && !(this.props.channelIsArchived || post.system_post_ids)) {
+        if (this.state.alt && !(this.props.channelIsArchived)) {
```
Comment (mickmister) [inline components/post_view/post/post.tsx:297]: Same question here about `post.system_post_ids`

File: components/post_view/post/post.tsx [hunk 10]
```diff
@@ -339,9 +350,9 @@ class Post extends React.PureComponent {
         if (!hideProfilePicture) {
             profilePic = (
                 <PostProfilePicture
-                    compactDisplay={this.props.compactDisplay}
-                    post={post}
+                    compactDisplay={Boolean(this.props.compactDisplay)}
                     userId={post.user_id}
+                    post={post}
```
Comment (mickmister) [inline components/post_view/post/post.tsx:355]: Can we move `post` back above `userId` to minimize the diff?

File: components/post_view/post/post.tsx [hunk 11]
```diff
@@ -360,16 +371,15 @@ class Post extends React.PureComponent {
         }
 
         return (
-            <PostContext.Provider value={{handlePopupOpened: this.handleDropdownOpened}}>
+            <PostContext.Provider value={{handlePopupOpened: this.handleDropdownOpened} as any}>
```
Comment (mickmister) [inline components/post_view/post/post.tsx:374]: Do we have to use `as any` here?

File: components/post_view/post_body/post_body.jsx [hunk 12]
```diff
@@ -71,7 +71,7 @@ export default class PostBody extends React.PureComponent {
         /**
          * Whether or not the post username can be overridden.
          */
-        enablePostUsernameOverride: PropTypes.bool.isRequired,
+        enablePostUsernameOverride: PropTypes.bool,
```
Comment (mickmister) [inline components/post_view/post_body/post_body.jsx:74]: Why is this no longer required?

File: components/post_view/post_header/post_header.jsx [hunk 13]
```diff
@@ -63,17 +63,17 @@ export default class PostHeader extends React.PureComponent {
...
          * If the user that made the post is a bot.
          */
-        isBot: PropTypes.bool.isRequired,
+        isBot: PropTypes.bool,
 
         /**
          * If the user that made the post is a guest.
          */
-        isGuest: PropTypes.bool.isRequired,
+        isGuest: PropTypes.bool,
```
Comment (mickmister) [inline components/post_view/post_header/post_header.jsx:76]: Why are these no longer required?

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @dudupopkhadze :+1: Just a few requests and questions

**Comment by jfrerich** [comment]
Hi, @dudupopkhadze! Could you please fix the conflicting files and address the suggestions from @mickmister?  Thanks for your contribution!

**Pull request closed without merging**